Add Firewall Settings UI

nateweller opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

This PR adds a new `firewall` module to Jetpack, and related settings UI. This also adds a check to the WAF `run.php` file to only run when the module is enabled.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Adds a new `firewall` Jetpack module
* Adds a new class to the WAF package, that registers a new `jetpack/v4/waf` endpoint which provides WAF settings data from the server (`bootstrapPath` and `hasRulesAccess`).
* Adds `<Firewall />` settings component that outputs all of the new settings UI.
* Updates the WAF package `rules.php` to abort execution if the `firewall` module is disabled, and adds a method for providing the path to `bootstrap.php`.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

1201069996155217-as-1202025817736945/f

p1HpG7-ffB-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->
* Boot up a Jetpack site running this branch and visit `/wp-admin/admin.php?page=jetpack#/settings`

**UI**
* Validate UI against the designs found in p1HpG7-ffB-p2.

**Module Toggling**
* Validate that the module is enabled by default.
* Generate the `bootstrap.php` from the CLI and validate that the firewall runs by default.
* Disable the module and validate that the firewall does not run.
* Validate that all setting toggles and text boxes save and load correctly.

**Upgrade Banner**
* Validate that the banner is not displayed on a site containing a plan with Jetpack Scan.
* Validate that the banner is displayed on a site not containing a plan with Jetpack Scan.
* Validate that the upgrade button takes you to the Scan/Security upgrade page.

**Enhance Protection Info**
* Validate that the path to `bootstrap.php` is correct.

#### Screenshots:

Default with no Scan plan:

<img width="780" alt="Screen Shot 2022-04-14 at 10 33 50 AM" src="https://user-images.githubusercontent.com/10933065/163434341-2000225f-64c6-41ee-85fd-2d922311fd93.png">

All settings enabled with Scan plan:

<img width="776" alt="Screen Shot 2022-04-14 at 10 33 04 AM" src="https://user-images.githubusercontent.com/10933065/163434512-ded3e9a1-ffa8-4d82-b6e7-761ba638b972.png">

All settings disabled:

<img width="781" alt="Screen Shot 2022-04-14 at 10 31 55 AM" src="https://user-images.githubusercontent.com/10933065/163434591-0436f543-86ea-44b8-aa4b-e59cc87ef005.png">

nateweller commented on projects/plugins/jetpack/modules/firewall.php:
```diff
@@ -0,0 +1,90 @@
+/**
+ * Module Name: Firewall
+ * Module Description: Protect your site with Jetpack's Web Application Firewall
+ * Sort Order: 5
```
Any framework of rhyme or reason I should be considering when picking this number?

nateweller commented on projects/plugins/jetpack/modules/firewall.php:
```diff
@@ -0,0 +1,90 @@
+<?php // phpcs:ignore WordPress.Files.FileName.InvalidClassFileName
+/**
+ * Module Name: Firewall
```
Do we want to be naming things as "WAF" and not "Firewall" everywhere in the codebase? I chose "Firewall" here as that is the actual title of the module that the user interacts with.

ArSn replied:
Wouldn't it be easier if we called it "WAF" everywhere? What would be the distinction between when we call it WAF and when we call it firewall? At least on the technical side, it has been WAF everywhere that I am aware of so far.

nateweller commented:
("All commits were linted before commit" is just due to reverts)

singerb commented:
I'm trying to test this but running into issues. However, so far the PHP side of this looks good. @miguelxpn or @bhoop can you take a look at the React side?

miguelxpn commented on projects/plugins/jetpack/_inc/client/security/firewall.jsx:
```diff
@@ -0,0 +1,305 @@
+	 */
+	state = {
+		jetpack_firewall_ip_list: this.props.getOptionValue( 'jetpack_firewall_ip_list' ),
+		jetpack_firewall_ip_allow_list: this.props.getOptionValue( 'jetpack_firewall_ip_allow_list' ),
```
Would you mind changing the option names to `jetpack_waf_ip_allow_list` and `jetpack_waf_ip_block_list`? After talking with @ArSn while looking at the backend site he reminded that we already have some options that start with jetpack_waf, so having some that starts with jetpack_firewall would be a bit odd. Sorry about asking this after agreeing with the naming. 😅

nateweller replied:
Updated!

ArSn commented on projects/plugins/jetpack/_inc/client/security/firewall.jsx:
```diff
@@ -0,0 +1,305 @@
+									] )
+								}
+								name="jetpack_firewall_ip_block_list"
+								placeholder={ __( 'Example: 12.12.12.1-12.12.12.100', 'jetpack' ) }
```
as of now, the IP checks do not support ranges. We are expecting an array of single IP addresses, so it would probably make sense to reflect that somehow here (e.g. either 1 IP per line or comma separated). Ranges and CIDR blocks will certainly come in a future version.

nateweller replied:
I've updated the placeholder here to display a list of single IP addresses :+1:

ArSn commented on projects/plugins/jetpack/_inc/client/security/firewall.jsx:
```diff
@@ -0,0 +1,305 @@
+									] )
+								}
+								name="jetpack_firewall_ip_allow_list"
+								placeholder={ __( 'Example: 12.12.12.1-12.12.12.100', 'jetpack' ) }
```
same thing here about the IP ranges

nateweller replied:
I've updated the placeholder here to display a list of single IP addresses :+1:

ArSn commented:
Merged master into this branch to have the IP block list and allow list server side functionality in this branch in order to base the feature-interconnectivity work on this branch while the PR is still in review. See 1201069996155224-as-1202148145416678

ArSn commented:
After clarifying the changes mentioned in my last comment further - see p1650458364845089-slack-CBG1CP4EN and p1650460731138599-slack-CBG1CP4EN - I actually like the solution to ignore the sanitation check better since the sanitation does not make much sense to begin with in this particular instance. I merged the master in here again, resolving the conflict and having the linting now safe on all sides. Sorry again for the unintended side effects of my original merge 😬

nateweller commented:
Addressed the following:
* Standardized naming to use "waf" instead of "firewall" in the code, while still presenting to user as "Firewall". 
* Updated IP list placeholders.

ArSn reviewed (dismissed):
I can really only speak to the PHP part, but this one LGTM as well. Once the review from @miguelxpn is in, I think this is good to merge.

miguelxpn reviewed (dismissed):
The frontend stuff looks good to me! The only thing we need to figure out is the whole module activation issue.

ArSn commented:
Merged master again to get the PR check green.

dereksmart commented on projects/packages/waf/src/class-waf-runner.php:
```diff
@@ -65,6 +65,19 @@ public static function is_allowed_mode( $option ) {
+	 */
+	public static function is_enabled() {
+		if ( class_exists( 'Jetpack' ) ) {
+			return \Jetpack::is_module_active( 'waf' );
```
This `Jetpack` plugin method is only a wrapper for the `Modules()->is_active()` [package method](https://github.com/Automattic/jetpack/blob/master/projects/packages/status/src/class-modules.php#L25). Does it make sense to use that now to avoid refactoring when we bring it into the Protect plugin (assuming we are)?

nateweller replied:
Removed the usage of the `Jetpack` method via https://github.com/Automattic/jetpack/pull/23769/commits/5b4c7a07b1970d2da98c4b0d61ba047b60a3d717 :+1:

ArSn reviewed (dismissed):
Still LGTM.

kraftbj commented:
@nateweller any other changes pending for this one?

nateweller commented:
Based on p1650556975439039-slack-CBG1CP4EN, I've moved the endpoints into the WAF package. This will not only be useful when we need these for the Protect plugin, but also fixes an issue where the UI would not be displayed when the module was disabled, due to the `uiEnabled` value coming from the module. Now that it's in the package, this will not be a problem.

@kraftbj This is the last of the changes that I have planned!

kraftbj commented:
@nateweller Sounds good. Let me give it a last check and I'll approve it right away.

dereksmart reviewed (changes requested):
Gave it a quick spin and noticed a some behaviors that don't look intended: 
1. When the module is _inactive_, it does not show in the main settings at all. 
2. When the `JETPACK_WAF_UI` is _not_ set, you can still see and "activate" it from two places: 
The first is in `/wp-admin/admin.php?page=jetpack_modules`
The other is in wp cli, via `wp jetpack module list`, and you can activate it via `wp jetpack module activate waf`

nateweller commented:
@dereksmart Thanks for looking at this! 

#1 should be addressed in https://github.com/Automattic/jetpack/pull/23769/commits/9ba215cc00c8a23dcc81800322063877f38c2b54, which I think I may have pushed while you were reviewing? 

I've prompted some discussion with the team for #2, it's been mentioned that we may potentially remove the define sooner than later which would negate this one. Otherwise, I will have to look into how we could conditionally make the module available based on the constant.

kraftbj commented:
The blocker I see:

Without the UI flag, the module is shown as enabled already in the modules list and is actually enabled (e.g. I'm seeing `Closure on line 39 of wp-content/plugins/jetpack-dev/jetpack_vendor/automattic/jetpack-waf/actions.php` being fired on `plugin_loaded`.

To merge this, we should have it off until the gate is removed and/or have the gate also short-circuit the module itself if it is not activate (related to Derek's point).

nateweller commented:
Per some discussion (p1650575200897379-slack-C029WFNV69M) and our release plan (pbuNQi-2Oa-p2), I've gone ahead and removed the `JETPACK_WAF_UI` gate completely as we are now planning to include the UI in the public beta.

kraftbj commented on projects/plugins/jetpack/modules/waf.php:
```diff
@@ -0,0 +1,13 @@
+ * Sort Order: 5
+ * First Introduced: 10.9
+ * Requires Connection: No
+ * Auto Activate: Yes
```
Per the release plan, shouldn't this be No? pbuNQi-2Oa-p2

```suggestion
 * Auto Activate: No
```

kraftbj commented on projects/plugins/jetpack/modules/waf.php:
```diff
@@ -0,0 +1,13 @@
+ * Module Description: Protect your site with Jetpack's Web Application Firewall
+ * Sort Order: 5
+ * First Introduced: 10.9
+ * Requires Connection: No
```
This requires a connection since we get the rules from the API.

```suggestion
 * Requires Connection: Yes
```

kraftbj reviewed (commented):
I need to go through this with a bit finer-tooth comb and near the end of my day. Very immediately, I think we need to declare this as a connected module and not auto-activated since the intent is this to go beta for 10.9.

miguelxpn reviewed (commented):
Tested it out and it works great! Now we can see the settings even with the module disabled. Once we address the things @kraftbj pointed out we should be good to go!

ArSn commented on projects/plugins/jetpack/_inc/client/security/waf.jsx:
```diff
@@ -0,0 +1,299 @@
+						</li>
+						<li>
+							{ __(
+								"Don't forget to undo this action when Firewall is turned off, or you uninstall Jetpack.",
```
I just read that in the interface and was like "What? If I don't undo this when the firewall is turned off, I will uninstall Jetpack?" But what it actually means is that I should undo the action in either of those two cases, which does make sense :)

Therefore, small improvement suggestion.
```suggestion
								"Don't forget to undo this action when Firewall is turned off, or when you uninstall Jetpack.",
```

kraftbj replied:
That's a good point. Do you think we should have more proactive notices if someone deactivates the feature while leaving that in place or uninstall Jetpack?

admin notice if disabled with it still active?
e-mail to site admin if still active when Jetpack is deactivated/uninstalled?

ArSn replied:
@kraftbj Actually, for the case when the firewall is turned off, we can be smart enough about it that the firewall does not actually do anything (this is not yet happening but should be easy to do, see p1650568832851519/1650548147.844039-slack-C029WFNV69M). For the other case I am not sure how exactly the deactivation/uninstall routine looks. If we do indeed remove all files, making this more prominent sounds like a good idea. Or at least create an empty file that does nothing in its place, if possible.

ArSn reviewed (approved):
Only one mini nitpick in addition to what has already been said by the others.

singerb commented:
As the tags indicate, based on pbuNQi-2Oa-p2#comment-4568 don't merge this until after the 10.9 freeze is lifted.

jeherve commented on projects/packages/waf/src/class-waf-endpoints.php:
```diff
@@ -0,0 +1,80 @@
+	/**
+	 * WAF Endpoint Permissions Callback
+	 *
+	 * @return bool|WP_Error True if user can view the Jetpack admin page.
+	 */
+	public static function waf_permissions_callback() {
+		if ( current_user_can( 'jetpack_admin_page' ) ) {
+			return true;
+		}
+
+		return new WP_Error(
+			'invalid_user_permission_view_admin',
+			REST_Connector::get_user_permissions_error_msg(),
+			array( 'status' => rest_authorization_required_code() )
+		);
+	}
```
Should we limit this to site admins instead? Non-admins can currently access the Jetpack dashboard to manage other settings, but I don't think non-admins need to see or update WAF settings. Maybe `jetpack_manage_modules` would be more appropriate here?

nateweller commented on projects/plugins/jetpack/_inc/client/security/waf.jsx:
```diff
@@ -0,0 +1,299 @@
+					<hr />
+					<a href="https://jetpack.com/support/privacy/" target="_blank" rel="noopener noreferrer">
+						{ __( 'Privacy information', 'jetpack' ) }
+					</a>
```
Thanks for your review Jeremy! This one has been addressed via https://github.com/Automattic/jetpack/pull/23769/commits/b9a426dee6c272373aff204e196e503791b85309.

nateweller commented on projects/plugins/jetpack/_inc/client/security/waf.jsx:
```diff
@@ -0,0 +1,299 @@
+									] )
+								}
+								name="jetpack_waf_ip_allow_list"
+								placeholder={ __( 'Example:', 'jetpack' ) + '\n12.12.12.1\n12.12.12.2' }
```
Addressed via https://github.com/Automattic/jetpack/pull/23769/commits/67a2322e661e4463314cde4b0be25198ea2a5bd8 - used `sprintf` for the IP list with a note for translators, and merged the above privacy strings into a single string with the existing `createInterpolateElement` implementation.

kraftbj reviewed (approved):
Let's give this a go.

singerb commented on projects/packages/waf/src/class-waf-endpoints.php:
```diff
@@ -0,0 +1,80 @@
+	 */
+	private static function has_rules_access() {
+		// any site with Jetpack Scan can download new WAF rules
+		return \Jetpack_Plan::supports( 'scan' );
```
I think it's only used in the UI, so we'll be fine regardless. We're not trying to gate rules at runtime based on purchases, just in the API and giving people helpful notices in the UI.
